crates/cubecl-opt/src/passes/integer_range_analysis.rs

```diff
@@ -0,0 +1,263 @@
+use std::ops::{Add, Mul, Sub};
+
+use cubecl_core::ir::{ConstantScalarValue, Elem, Operation, Operator, Variable};
+
+use crate::{AtomicCounter, Optimizer, Range};
+
+use super::OptimizationPass;
+
+impl Range {
+    fn constant(val: i64) -> Self {
+        Self {
+            lower_bound: Some(val),
+            upper_bound: Some(val),
+        }
+    }
+
+    fn uint(upper: i64) -> Self {
+        Self {
+            lower_bound: Some(0),
+            upper_bound: Some(upper),
+        }
+    }
+}
+
+#[derive(Default, Clone, Debug)]
+pub struct IntegerRangeAnalysis;
+
+impl OptimizationPass for IntegerRangeAnalysis {
```

Comment:
It doesn't feel right that an analysis implements an optimization pass. I would expect two traits, one for analysis and one for the optimization. Optimizations might use the results of the analyses stored in the optimizer's state. 

```
Step 1 => Initialize the optimizer
Step 2 => Run all analyses and store their results in the optimizer's state.
Step 3 => Run all optimization pass that can access all analyses previously computed.
Step 4 => Generate or return the transformed CubeCL IR.
Step 5 => Enjor speed 😎
```

It might be the current workflow, it's just not named that way.

Reply:
The reason they're combined is because some analyses should run after certain optimizations but before certain other optimizations that depend on them. For example, integer analysis is greatly simplified by first doing assignment merging and constant evaluation, but must be done before in bounds transformation. Maybe I can just rename the trait to `OptimizerPass` to make it a bit more clear it's just anything the optimizer executes as part of its optimization loop.